### PR TITLE
Correct TextBox.PlaceholderText font transparency

### LINKED ISSUE #3642: TextBox.PlaceholderText uses BackColor even when TextBox.Enabled = false
* .NET Core Version: 3.1.6
* Have you experienced this same bug with .NET Framework?: No (PlaceholderText not available)
* Visual Studio version: 2019 16.7 Preview 5

**Problem description:**

When PlaceholderText is set and displayed, it's always drawn with the BackColor as the background, regardless of the Enabled value. This causes minor graphical issues as presented below.

image

The issue is most likely [here](https://github.com/dotnet/winforms/blob/b12ba3357e6f2246b5af3c506b8b6e05b8ce1cae/src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs#L900) - renderer for the Text itself [draws the appropriate background color first, then the text without any background](https://github.com/dotnet/winforms/blob/b12ba3357e6f2246b5af3c506b8b6e05b8ce1cae/src/System.Windows.Forms/src/System/Windows/Forms/TextBoxRenderer.cs#L91).

**Expected behavior:**

When PlaceholderText is set and TextBox is disabled, the background color is rendered properly across the entire control.

**Minimal repro:**

```csharp
using System;
using System.Drawing;
using System.Windows.Forms;

namespace BrokenPlaceholder {
    class PlaceholderForm : Form {
        private TextBox textBox1;
        private TextBox textBox2;
        private TextBox textBox3;

        public PlaceholderForm() {
            SuspendLayout();

            textBox1 = new TextBox {
                Name = "textBox1",
                Size = new Size(226, 23),
                Location = new Point(33, 25),
                Text = "Some perfectly fine Text"
            };

            textBox2 = new TextBox {
                Name = "textBox2",
                Size = new Size(226, 23),
                Location = new Point(33, 54),
                PlaceholderText = "PlaceholderText"
            };

            textBox3 = new TextBox {
                Name = "textBox3",
                Size = new Size(226, 23),
                Location = new Point(33, 83),
                PlaceholderText = "Disabled PlaceholderText",
                Enabled = false
            };

            ClientSize = new Size(293, 138);
            Controls.Add(this.textBox1);
            Controls.Add(this.textBox2);
            Controls.Add(this.textBox3);
            Name = "PlaceholderForm";
            Text = "PlaceholderForm";
            ResumeLayout(false);

            PerformLayout();
        }
    }

    static class Program {
        [STAThread]
        static void Main() {
            Application.SetHighDpiMode(HighDpiMode.SystemAware);
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new PlaceholderForm());
        }
    }
}
```

### PR DESCRIPTION
**Pull request by RussKie**
## Proposed changes

Whenever the `PlaceHolderText` property was set on a disabled textbox the placeholder text would be drawn with an opaque background.
Correct the text background transparency.

Fixes #3642.
Closes https://github.com/dotnet/winforms/pull/4008 as obsolete

## Screenshots 

Copied from https://github.com/dotnet/winforms/pull/4008

### Before

before

### After

after

## Test methodology 

- manual
- tests

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/4057)

### CODE REVIEW

File: src/System.Windows.Forms/tests/UnitTests/TextBoxTests.Rendering.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            {
+                Size = new Size(80, 23),
+                PlaceholderText = "abc",
+                Enabled = false
+            };
+            form.Controls.Add(textBox);
+
+            // Force the handle creation
+            Assert.NotEqual(IntPtr.Zero, form.Handle);
+            Assert.NotEqual(IntPtr.Zero, textBox.Handle);
```
Comment (hughbe) [inline src/System.Windows.Forms/tests/UnitTests/TextBoxTests.Rendering.cs:30]: nit/FYI: creating the parent handle creates the child controls

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 2]
```diff
@@ -969,10 +969,8 @@ protected override void WndProc(ref Message m)
 
             if (ShouldRenderPlaceHolderText(m))
             {
-                using (Graphics g = CreateGraphics())
-                {
-                    DrawPlaceholderText(g);
-                }
+                using var controlDC = new User32.GetDcScope(Handle);
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs:972]: Looked a little closer at this. We should be using `BeginPaint` if the message is `WM.PAINT` as the `HDC` in the message won't always be the one associated with the `Handle`. One concrete example of this is `WM.PRINT`, which turns around and calls `WM.PAINT` with the `HDC` it was given- and we use that functionality ourselves when grabbing a metafile.

The same problem existed with the existing code, as the `Graphics` should have been created from the passed in `HDC` when getting the `WM.PAINT` message. We can create a new issue and make a separate PR for clarity here if you want.

File: src/System.Windows.Forms/tests/UnitTests/TextBoxTests.Rendering.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            DeviceContextState state = new DeviceContextState(emf);
+
+            textBox.TestAccessor().Dynamic.DrawPlaceholderText(emf.HDC);
+
+            emf.Validate(
+                state,
+                Validate.TextOut(
+                    "abc",
+                    bounds: null,                                   // Don't care about the bounds for this test
+                    State.BackgroundMode(Gdi32.BKMODE.TRANSPARENT)));
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/tests/UnitTests/TextBoxTests.Rendering.cs:42]: Could you add this state validator back to `TextRenderer_DrawText_DefaultBackground_RendersTransparent`?

Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/TextBoxTests.Rendering.cs:42]: done

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: The current implementation draws the placeholder text only in `WM_PAINT`, so the handle is valid:
https://github.com/dotnet/winforms/blob/6391b524dd64e7530eba986c6607bf129b3ec8ae/src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs#L979-L984

But I noticed a flickering in the current implemenations:
placeholdertext-current

Then I tried changing the code as follows:
```diff
diff --git a/src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs b/src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs
index 201533cd7..5430f6737 100644
--- a/src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs
+++ b/src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs
@@ -959,19 +959,24 @@ namespace System.Windows.Forms
                         base.WndProc(ref m);
                     }
                     break;
+
+                case User32.WM.PAINT:
+                    if (ShouldRenderPlaceHolderText(m))
+                    {
+                        using var hdc = new User32.BeginPaintScope(Handle);
+                        DrawPlaceholderText(hdc);
+                    }
+                    else
+                    {
+                        base.WndProc(ref m);
+                    }
+                    break;
+
                 case User32.WM.PRINT:
                     WmPrint(ref m);
                     break;
                 default:
                     base.WndProc(ref m);
                     break;
             }
-
-            if (ShouldRenderPlaceHolderText(m))
-            {
-                using var controlDC = new User32.GetDcScope(Handle);
-                DrawPlaceholderText(controlDC.HDC);
-            }
         }
```

...and the strange artifact only got worse - the placeholder text is getting drawn multiple times when the textbox is getting hovered over and over again:
placeholdertext-pain

Apparently when the mouse is moved hovering the textbox multiple `WM_PAINT` messages are sent, and we have a re-entrancy issue. But I'm unsure how to best solve it. I tried tracking mouse enter/leave, but ultimately I failed to find a solution.
Calling for help @JeremyKuhne @weltkante.

Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: Reentrancy should not happen, you shouldn't get `WM_PAINT` in the middle of a `WM_PAINT` message processing, the UI thread is synchronous so it can only happen if you do message pumping in `WM_PAINT`, which is a bad thing to do. If you have reentrancy you need to look at the stack trace of your reentrant call to understand what you did that triggered message pumping, you are most likely doing something wrong. (Some .NET infrastructure like locking primitives may do message pumping and need special workarounds, so if you need locking in code called from `WM_PAINT` you may require those workarounds to suppress message pumping!)

Receiving multiple `WM_PAINT` _in sequence_ is pretty normal, there is a special flag fErase in the [`PAINTSTRUCT`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-paintstruct) which tells `WM_PAINT` if the background needs to be cleared, when set `BeginPaint` will raise `WM_ERASEBKGND`. Normally everything should work out.

However overriding `WM_PAINT` without calling the base WndProc is using undocumented behavior (and differ with different Windows versions) because you need to rely on knowledge how the native control is implemented. Maybe `WM_ERASEBKGND` of the native textbox is a no-op and clearing the background happens in `WM_PAINT` of the native textbox? Then suppressing the `WM_PAINT` will also suppress correct background rendering.

I can take a closer look and debug this.

Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: I've debugged this a bit and think the whole placeholder text thing is inherently flawed. `WM_PAINT` does not seem extensible by subclassing.

> I noticed two things in the original implementation - a flickering and double text write

The native textbox knows exactly what is dirty and needs to be refreshed, the placeholder text is interfering with the assumptions of the native implementation. The double text writes happen because the native WndProc "knows" there is no text so it doesn't clear the background, only updates the frame.

> Then I tried changing the code as follows [...] and the strange artifact gets somewhat "better" - no more flickering, but the placeholder text is still getting drawn multiple times when the textbox is getting hovered over and over again

- you suppressed the native WM_PAINT handler, so obviously the interference between your code and the native TextBox rendering is gone
- the double text rendering is still happening because now nobody is drawing the background. The native TextBox seems to implement `WM_ERASEBKGND` as a noop (an optimization because it knows that the visual style animations will render everything opaquely in WM_PAINT).

> Apparently when the mouse is moved hovering the textbox multiple WM_PAINT messages are sent

This is the animation of the border color and to be expected.

> and we have a re-entrancy issue

I could not reproduce any reentrancy

---

Code which works for me:

```csharp

// Invalidate the whole control to make sure the native control doesn't make any assumptions over what it has to paint
if (ShouldRenderPlaceHolderText(m))
{
    Invalidate();
}

// Let the native implementation draw the background and animate the frame
base.WndProc(ref m);

if (ShouldRenderPlaceHolderText(m))
{
    // Invalidate again because the native WM_PAINT already validated everything by calling BeginPaint itself.
    Invalidate();

    // Use BeginPaint instead of GetDC to prevent flicker.
    using var hdc = new User32.BeginPaintScope(Handle);

    DrawPlaceholderText(hdc);
}
```

Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: Thanks for the investigation @weltkante! Your code above looks right to me.

I'm unclear on why we re-painted the placeholder text on `WM.KILLFOCUS`. If we really needed to do so we should just invalidate in that case, right?

Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: Yes, the `WM_KILLFOCUS` check in `ShouldRenderPlaceHolderText` is dead now and needs investigation why it was necessary. If its necessary it should be replaced by invalidation instead of immediate repaint, thats always better because Windows can aggregate multiple invalidations, while doing immediate repaints means you have to pay for every repaint.

For example program logic may move the focus around multiple times before it settles, invalidation means there is only one repaint at the end, while immediate repaint means the focus logic becomes orders of magnitude more expensive.

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: Thank you @weltkante!

"Re-entrancy" was a bad choice of word, at the point when I've submitted the message I think I was on a 5th or 7th iteration of it while trying different approaches...

I don't think `WM_KILLFOCUS` is necessary since the method is invoked only on `WM_PAINT`. At least in my tests its presence or absence didn't make any difference.

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: I tried the proposal, but it still flickers (With or without `WM_KILLFOCUS`):
placeholdertext-new

**Comment by RussKie** [comment]
Ready for review.

We had a long debugging session with @JeremyKuhne, and it seems the flickering issue is more complex than originally anticipated. The native control (or something else) keeps on invalidating the control that queues up `WM_PAINT` event.
I have restructured the code as per the @weltkante's and @JeremyKuhne's suggestions factoring out `CreateGraphics` call, and moving the handling into `WM_PAINT`.

Sadly besides the flicker, while testing we discovered more (rather concerning) issues with the current implementations - notably total unawareness of placeholder text in the native control, if once set; and UIA gaps. I have noted the issues on the original implementation in https://github.com/dotnet/winforms/pull/96#issuecomment-705355778, and these issues will be dealt with separately.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs (hunk 2 above)
Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: Yeah, I thought that would happen, but since I didn't see it in my test I wasn't commenting further on the problems beyond saying this approach is inherently flawed. (Doesn't seem to happen for me if you move the mouse slowly and let the animation finish. Or maybe thats just my machine being fast enough when the CPU is not under stress.)

I'm guessing the problem is that there are two BeginPaint/EndPaint calls so Windows may have a chance to present to screen between the two calls, after the background is cleared but before the placeholder text is drawn.

Its a bit unclear to me what the rules are, Windows has to have some internal locking when it doesn't present to screen, otherwise everything would flicker all the time if things could get presented in the middle of painting. (Every control needs to clear its background before painting on it, so if Windows could present in between everything would randomly flicker.)

What I hoped was that it wouldn't present in the middle of a WM_PAINT call if it properly painted via BeginPaint, but apparently thats not always true. Maybe the first EndPaint signals that "WM_PAINT is done and can be presented" before it actually returns? Then you'd have a race condition which flickers if the rest of WM_PAINT (your placeholder text drawing) doesn't render quickly enough.

In summary I don't think this is fixable, at least not without some arcane workaround and someone who knows the exact rules of how WM_PAINT presentation works. Like I said initially the idea of extending WM_PAINT of a native control by drawing over it is inherently flawed.

Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: I'm wondering if I'm missing something, [docs](https://docs.microsoft.com/en-us/windows/win32/controls/user-controls-intro#subclassing-the-window-class-of-an-existing-control) explicitly spell out that this is supposed to be supported subclassing and gives overriding WM_PAINT as an example (unfortunately without any sample code how to do it without flicker)

> Subclassing an existing control is another way to create a custom control. The subclass procedure can alter selected behaviors of the control by processing those messages that affect the selected behaviors. All other messages pass to the original window procedure for the control. For example, an application can display a small bitmap next to the text in a read-only, single-line edit control by subclassing the control and processing the WM_PAINT message. For more information, see About Window Procedures and Subclassing Controls.

This would be exactly what you are trying to do, I'm wondering what is breaking the scenario.

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: If we do all the drawing ourselves (i.e. `UserPaint`), then it probably won't happen. But we're attempting to do custom paint in non-user paint mode, perhaps this is the real issue here.
We've gone through some of the EditControl code (not exhaustively due to time restrictions), but there wasn't anything that jumped out.

Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: > If we do all the drawing ourselves (i.e. UserPaint), then it probably won't happen.

Well, definitely, thats a well defined extension point provided by the EditControl, so it should be aware. I think you only need to do the text painting, the user painted EditControl is documented to still draw the frame if I understood the doc correctly. Maybe you could try switching to UserPaint when `ShouldRenderPlaceHolderText` is true (that is, text is empty and placeholder text is present).

On the other hand, if nothing prevents usage of `EM_SETCUEBANNER` then thats probably the sane way to do it. Wasn't aware of this since this feature/message is not mentioned on the EditControl overview doc, only in the reference documentation of the individual window messages.

> We've gone through some of the EditControl code (not exhaustively due to time restrictions), but there wasn't anything that jumped out.

I don't think its an issue with the EditControl, everything is behaving as I'd expect from a normal win32 native control. Its just that the API and the control is not designed to be extensible this way. Its a more fundamental thing how `WM_PAINT` and `BeginPaint`/`EndPaint` works, so it would require looking in the user32/gdi32 codebase. To correctly work around the problem it needs understanding of the synchronization with the composition manager which already prevents presentation at arbitrary points in time during painting, preventing flicker under conditions which are not entirely obvious, as I explained above.

However, thats probably an investigation not worth the effort.

Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: > However, thats probably an investigation not worth the effort.

Yeah, although I _wish_ I had the time to dig into it further. We're getting a second WM_PAINT after we do our begin/end paint. Something is apparently invalidating the control again. Explicitly calling `ValidateRect` should, in theory, prevent this second call but it clearly isn't happening.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 4]
```diff
@@ -936,15 +936,15 @@ private void DrawPlaceholderText(Graphics graphics)
                 }
             }
 
-            TextRenderer.DrawText(graphics, PlaceholderText, Font, rectangle, SystemColors.GrayText, BackColor, flags);
+            TextRenderer.DrawTextInternal(hdc, PlaceholderText, Font, rectangle, SystemColors.GrayText, Gdi32.QUALITY.DEFAULT, flags);
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs:939]: We should expose the `TextRenderer.s_defaultQuality` as we're switching from `Graphics`.

**Review by JeremyKuhne — CHANGES_REQUESTED** [review]
